File: sliding_window_maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        int left = 0;
        int right = 0;
        deque<int> q;

        while (right < nums.size()) {
            while (!q.empty() && nums[right] > nums[q.back()]) {
                q.pop_back();
            }
            q.push_back(right);

            if (left > q.front()) {
                q.pop_front();
            }

            if (right + 1 >= k) {
                res.push_back(nums[q.front()]);
                left++;
            }
            right++;
        }

        return res;        
    }
};

File: sliding_window_maximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function(nums, k) {
    const res = [];
    let left = 0;
    let right = 0;
    const q = [];

    while (right < nums.length) {
        while (q.length > 0 && nums[right] > nums[q[q.length - 1]]) {
            q.pop();
        }
        q.push(right);

        if (left > q[0]) {
            q.shift();
        }

        if (right + 1 >= k) {
            res.push(nums[q[0]]);
            left++;
        }
        right++;
    }

    return res;    
};
